### ISSUE TITLE
mac OS gamepad does not map SDL properly

### ISSUE THREAD

**Issue by salvob41** [godotengine/godot#26544]
**Godot version:**
All godot 3.x

**OS/device including version:**
macOS 

**Issue description:**
Mappings in SDL format do not match up with Godot Buttons (bottom, right, up and left). 

I thought this was just for a particular gamepad (described in this isssue #25138 ) but I actually tried with two other gamepad and the problem was still there. 

In windows the same gamepads works perfectly. 
This still stands even if I forcely put the mapping string with something like:
`Input.add_joy_mapping("*mapping_from_SDL")`

**Comment by bruvzg** [comment]
> This still stands even if I forcely put the mapping string

Try using `Input.add_joy_mapping("mapping string", true)` (with extra true argument), if default db already have invalid mapping for your controller this will overwrite it, without true it's only adding new records.

**Comment by salvob41** [comment]
yes, I did. Same results though

**Comment by salvob41** [comment]
I'll post here a relevant GIF that I used to explain in #25138 . Same thing happened for another, different, gamepad
![bug_mapping_macos](https://user-images.githubusercontent.com/6097015/52863882-3397e400-3139-11e9-806f-7418c2218ac8.gif)

**Comment by akien-mga** [comment]
CC @Hinsbart

**Comment by bruvzg** [comment]
Note that your controller mapping is not imported from SDL db, there's following error for it (and for multiple other controllers) in the `default_controller_mappings.gen.cpp`:

```
// WARNING - DISCARDED UNSUPPORTED MAPPING TYPE FROM DATABASE .../gamecontrollerdb.txt: Mac OS X 030000008f0e00000300000009010000,2In1 USB Joystick,+leftx:h0.2,+lefty:h0.4,-leftx:h0.8,-lefty:h0.1,a:b2,b:b1,back:b8,dpdown:+a1,dpleft:-a0,dpright:+a0,dpup:-a1,leftshoulder:b4,leftstick:b10,lefttrigger:b6,rightshoulder:b5,rightstick:b11,righttrigger:b7,rightx:a2,righty:a3,start:b9,x:b3,y:b0,platform:Mac OS X,
```

Most likely `Input.add_joy_mapping(...)` doesn't accept mapping copied from `gamepad-tool` for the same reasons.

SDL db had some changes in 2.0.6/2.0.5/2.0.4 version. Probably some are not supported by Godot.

**Comment by salvob41** [comment]
@bruvzg ! Thank you so much ! 
I was going to say that Input.add_joy_mapping(...) seems not to work with new mappings. But there is no error message. 
What do you recommend to do?

**Comment by salvob41** [comment]
and where did you see the error messages ?

**Comment by bruvzg** [comment]
>and where did you see the error messages ?

In the generated source file `main\default_controller_mappings.gen.cpp`, it is creating by following script: [main/main_builders.py](https://github.com/godotengine/godot/blob/master/main/main_builders.py#L101). Looks like this script do not support 2.0.6 range and inversion modifiers (+,-,~).

**Comment by bruvzg** [comment]
Try using older (1.1.2) version of the `gamepad-tool` to generate mapping in 2.0.5 format (http://www.generalarcade.com/gamepadtool/GamepadTool.osx.1.1.2.zip)

**Comment by salvob41** [comment]
> In the generated source file main\default_controller_mappings.gen.cpp, it is creating by following script: main/main_builders.py. Looks like this script do not support 2.0.6 range and inversion modifiers (+,-,~).

Should we open a separate issue about this, @bruvzg, @akien-mga   ?

**Comment by akien-mga** [comment]
> Should we open a separate issue about this, @bruvzg, @akien-mga ?

There's the old #8368 already which was never worked on. Might be worth adding what @bruvzg found out there and promote it to a bug.

**Comment by salvob41** [comment]
also might be good to add an error to output or somewhere when `Input.add_joy_mapping(...)` cannot add a mapping (or the mapping is somewhat invalid)

**Comment by akien-mga** [comment]
Can you check if it works fine in 3.2 beta 6 or later?
I might have fixed this with #34910.

**Comment by akien-mga** [comment]
I'll assume it fixed by #34910 for now, and if it isn't, it should be fixable by contributing the missing mappings to https://github.com/gabomdq/SDL_GameControllerDB.

Please comment if you can still reproduce the issue / need help to contribute mappings upstream if they're missing.
